llvm/lib/Target/AMDGPU/AMDGPUFixLiveRangePreWaveRA.cpp

```diff
@@ -0,0 +1,258 @@
+      if (PostDomMBB == MBB) {
+        if (auto *ParentNode = PDT->getNode(MBB)->getIDom())
+          PostDomMBB = ParentNode->getBlock();
+      }
+      // walk PDT from SuccMBB to PostDomMBB
+      // add MBB as the control-parent of the blocks along the path (except
+      // PostDomBB)
+      for (auto *Node = PDT->getNode(SuccMBB);
+           Node && Node->getBlock() != PostDomMBB; Node = Node->getIDom()) {
+        auto *PathMBB = Node->getBlock();
+        CtrlDeps[PathMBB].insert(MBB);
+      }
+    }
+  }
+}
+
+bool AMDGPUFixLiveRangePreWaveRA::influences(MachineBasicBlock *CtrlMBB,
+                                             MachineBasicBlock *DepMBB) {
+  if (CtrlDeps.find(DepMBB) == CtrlDeps.end())
+    return false;
+
+  SmallVector<MachineBasicBlock *, 8> WL;
+  SmallPtrSet<MachineBasicBlock *, 8> Visited;
+  for (auto *ParMBB : CtrlDeps[DepMBB]) {
+    WL.push_back(ParMBB);
+  }
+
+  while (!WL.empty()) {
+    auto *MBB = WL.back();
+    WL.pop_back();
+    Visited.insert(MBB);
+    if (MBB == CtrlMBB)
+      return true;
+    if (CtrlDeps.find(MBB) != CtrlDeps.end()) {
+      for (auto *ParMBB : CtrlDeps[DepMBB]) {
+        if (!Visited.count(ParMBB))
+          WL.push_back(ParMBB);
+      }
+    }
+  }
+
+  return false;
+}
+
+bool AMDGPUFixLiveRangePreWaveRA::runOnMachineFunction(MachineFunction &MF) {
+  LLVM_DEBUG(dbgs() << "AMDGPUFixLiveRangePreWaveRA: function " << MF.getName()
+                    << "\n");
+
+  const GCNSubtarget &ST = MF.getSubtarget<GCNSubtarget>();
+
+  TII = ST.getInstrInfo();
+  TRI = &TII->getRegisterInfo();
+  MRI = &MF.getRegInfo();
+  PDT = &getAnalysis<MachinePostDominatorTree>();
+
+  buildControlDependences(MF);
+
+  ReversePostOrderTraversal<MachineFunction *> RPOT(&MF);
+  bool Changed = false;
+  for (unsigned i = 0, e = TRI->getNumRegUnits(); i != e; ++i) {
+    for (MCRegUnitRootIterator Root(i, TRI); Root.isValid(); ++Root) {
+      auto RC = TRI->getPhysRegBaseClass(*Root);
+      if (!RC || TRI->isSGPRClass(RC)) {
+        // dbgs() << printReg(*Root, TRI) << "\n";
+        // skip reg-class that is not relevant
+        continue;
+      }
+      for (MCPhysReg Reg : TRI->superregs_inclusive(*Root)) {
+        // if a reg is either not-seen or reserved, not a concern for RA
+        if (MRI->reg_empty(Reg) || MRI->isReserved(Reg))
+          continue;
+
+        // iterate through the CFG, processing every divergent branch
+        for (MachineBasicBlock *MBB : RPOT) {
+          MachineBasicBlock *TrueMBB = nullptr;
+          MachineBasicBlock *FalseMBB = nullptr;
+          SmallVector<MachineOperand, 1> Cond;
+          TII->analyzeBranch(*MBB, TrueMBB, FalseMBB, Cond);
+
+          if (!Cond.size())
+            break;
+
+          auto CondOpnd = Cond.back();
+          if (!FalseMBB)
+            FalseMBB = MBB->getNextNode();
+
+          // check if this is a divergent branch
+          // is this the right way?
+          if (CondOpnd.getReg() != AMDGPU::VCC &&
```

Comment:
Yes this won't be sufficient.
I think the idea long term is to use MachineUniformityInfo.